pulsar-client/src/main/java/org/apache/pulsar/client/impl/ControlledClusterFailover.java

```diff
@@ -0,0 +1,235 @@
+
+                    Authentication authentication = null;
+                    if (!Strings.isNullOrEmpty(controlledConfiguration.authPluginClassName)
+                            && !Strings.isNullOrEmpty(controlledConfiguration.getAuthParamsString())) {
+                        authentication = AuthenticationFactory.create(controlledConfiguration.getAuthPluginClassName(),
+                                controlledConfiguration.getAuthParamsString());
+                    }
+
+                    String tlsTrustCertsFilePath = controlledConfiguration.getTlsTrustCertsFilePath();
+                    String serviceUrl = controlledConfiguration.getServiceUrl();
+
+                    if (authentication != null) {
+                        pulsarClient.updateAuthentication(authentication);
+                    }
+
+                    if (!Strings.isNullOrEmpty(tlsTrustCertsFilePath)) {
+                        pulsarClient.updateTlsTrustCertsFilePath(tlsTrustCertsFilePath);
+                    }
+
+                    pulsarClient.updateServiceUrl(serviceUrl);
+                    currentPulsarServiceUrl = serviceUrl;
+                    currentControlledConfiguration = controlledConfiguration;
+                }
+            } catch (IOException e) {
+                log.error("Failed to switch new Pulsar url, current: {}, new: {}",
+                        currentControlledConfiguration, controlledConfiguration, e);
+            }
+        }), interval, interval, TimeUnit.MILLISECONDS);
+    }
+
+    public String getCurrentPulsarServiceUrl() {
+        return this.currentPulsarServiceUrl;
+    }
+
+    public URL getPulsarUrlProvider() {
+        return this.pulsarUrlProvider;
+    }
+
+    protected ControlledConfiguration fetchControlledConfiguration() throws IOException {
+        // call the service to get service URL
+        InputStream inputStream = null;
+        try {
+            URLConnection conn = pulsarUrlProvider.openConnection();
+            inputStream = conn.getInputStream();
+            String jsonStr = new String(IOUtils.toByteArray(inputStream), StandardCharsets.UTF_8);
+            ObjectMapper objectMapper = getObjectMapper();
+            return objectMapper.readValue(jsonStr, ControlledConfiguration.class);
+        } catch (IOException e) {
+            log.warn("Failed to fetch controlled configuration. ", e);
+            return null;
+        } finally {
+            if (inputStream != null) {
+                inputStream.close();
+            }
+        }
+    }
+
+    private ObjectMapper getObjectMapper() {
+        if (objectMapper == null) {
+            objectMapper = new ObjectMapper();
+        }
+        return objectMapper;
+    }
+
+    @Data
+    protected static class ControlledConfiguration {
+        private String serviceUrl;
+        private String tlsTrustCertsFilePath;
+
+        private String authPluginClassName;
+        private String authParamsString;
```

Comment:
So, we need different URLs for multiple clients using the ControlledClusterFailover. 😭

Reply:
I add the `Map<String, String> header` configuration for http request header. User can set a token in header and the serviceUrl server parse the token and return related authentication configuration to support multiple clients.